### DISCUSSION TITLE
Nested pydantic request model raising ValueError returns 500 instead of 4xx

### DISCUSSION THREAD

**Discussion by larissaleite** [fastapi/fastapi#6686] [Questions] [answered]
First of all, thanks for the great work with FastAPI 🥇  It's very much appreciated!

I think the problem I have is more related to FastAPI than pydantic because I manage to get the intended behavior when using the same pydantic models outside of the FastAPI context.

### Example

I have something like:

```Python
from typing import List
import uvicorn
from fastapi import FastAPI
from pydantic import BaseModel, validator

app = FastAPI()


class Child(BaseModel):
    name: str

    @validator("name")
    def _validate_name(cls, value):
        raise ValueError("wrong name!")


class Parent(BaseModel):
    child: Child
    foo: str

    @validator("foo")
    def _validate_foo(cls, value, values):
        name = values["child"].name
        if value != name:
            raise ValueError("name should match")
        return value


class SubModel(BaseModel):
    parents: List[Parent]


class RequestPayload(BaseModel):
    payload: List[SubModel]


@app.post("/")
def root(payload: RequestPayload):
    return {"foo": "bar"}


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=9999)
```

### Description

When calling the above endpoint as follows: `curl -X POST -H "Content-Type: application/json" http://0.0.0.0:9999/ -d '{"payload": [{"parents": [{"child": {"name": "child"}, "foo": "bar"}]}]}'` I get `Internal Server Error` as a response, failing in the `validate_foo` with `KeyError: 'child'`. I expected the API to return 4xx and and the value error raised in the validation of the `Child` model. Shouldn't the error have stopped the validation altogether having the API return 4xx?
Am I doing something wrong?

Thanks for taking the time to read my question ❤️ 

### Environment

* OS: MacOS
* FastAPI Version: 0.65.2
* Pydantic Version: 1.8.2

**Comment by Kludex** [comment, marked as answer]
## Problem

The path of your data is this:
1. Data gets on the `root` endpoint.
2. `RequestPayload` :heavy_check_mark: 
3. `List[SubModel]` :heavy_check_mark: 
4. `Parent` :standing_person: 
5. Validate `child` :open_mouth: 
6. Find out that the field `name` of `Child` is invalid! :cry: 
7. Try to use `values["child"]`, but `values` doesn't have `child` because it was not valid. :wine_glass: 

## How to solve this issue

Always make sure that you have the key you want from `values`:
```python
from typing import List
import uvicorn
from fastapi import FastAPI
from pydantic import BaseModel, validator

app = FastAPI()


class Child(BaseModel):
    name: str

#    @validator("name")
#    def _validate_name(cls, value):
#        raise ValueError("wrong name!")


class Parent(BaseModel):
    child: Child
    foo: str

    @validator("foo")
    def _validate_foo(cls, value, values):
        # print(values)  # This print here can be valuable for debugging! :)
        if "child" in values and values["child"].name != value:
            raise ValueError("name should match")
        return value


class SubModel(BaseModel):
    parents: List[Parent]


class RequestPayload(BaseModel):
    payload: List[SubModel]


@app.post("/")
def root(payload: RequestPayload):
    return {"foo": "bar"}


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=9999)
```
You can check that all examples on [Pydantic documentation](https://pydantic-docs.helpmanual.io/usage/validators/) follow what I just recommended.

> Thanks for taking the time to read my question :heart: 

You're welcome! :smile:

**Comment by larissaleite** [comment]
Thanks for the answer 🙂  I will make the changes accordingly, but shouldn't the `ValueError` get propagated from `Child` all the way back to the request? Just trying to understand the behavior

**Comment by Kludex** [comment]
Well, if you don't get the KeyError, it will be. 🤔

**Comment by larissaleite** [comment]
Ah, that's correct 🙂  thanks for the quick reply, this is solved now ❤️
